torch/_dynamo/variables/builder.py

```diff
@@ -1233,7 +1237,11 @@ def _clone_input(value):
         example_value = _clone_input(example_value)
         proxy.node.meta["example_value"] = example_value
         specialized_props = target_cls.specialize(example_value)
-        if isinstance(example_value, torch._subclasses.fake_tensor.FakeTensor):
+        # TODO: not sure about this fake mode test
```

Comment:
I'm confused too. I thought mixed fake modes is not allowed? But it looks as if it is expected and no error raised if spotted.

Reply:
The original check would assume the fake tensor was created by the global fake mode, because dynamo was designed to instantiate a fake mode internally and use it, so it would never have a mismatch

Now, we allow fake modes to be called outside `torch._dynamo.export` too, so now this check was extended to make sure that if the user input is fake, the fake mode used to fakefy it matches the one configured for the `torch._dynamo.export` lifecycle.

This basically tightens the fake_mode checks within dynamo, it doesn't change any behavior

Reply:
Let's write it as check + raise when mixed fake mode is detected then?

Reply:
That is already done early (before FX tracing begins) in the execution path within `FakeTensorMode.dispatch` call stack, which includes a `validate_and_convert_non_fake_tensors(func, converter, args, kwargs)` step.

Reply:
@ezyang what do you think? should we check whether mixed fake mode happens here or leave it as is (just checking fake mode matches)?

Reply:
So, with the modifications from this PR, we have two modes but they should not mix.

The outer mode is the user defined fake tensor mode. Inside Dynamo, these should show up as ambient tensors that we want to fakeify. This mode typically does not have symbolic shapes.

The inner mode is Dynamo's administrative fake tensor mode. All tensors get converted into this mode before we do reasoning on it. This mode should have symbolic shapes if you're modeling dynamism.

`wrap_fx_proxy_cls` can see both types of fake tensor. It gets called to wrap preexisting tensors (outer mode). It also gets called recursively after Dynamo has already fakeified a result (inner mode).